### ISSUE TITLE
Need Host environment variable resolution to pass some information to a container

### ISSUE THREAD

**Issue by owaaa** [aws/amazon-ecs-agent#3]
Before moving to ECS, I need to be able to make sure that I can run consul in that environment. To run consul, I need to pass some Host environment variables as environment variables to the container. I would normally do this as `-e HOST=$HOST` etc. However, it seems that that if I define a variable with `$HOST` in a task definition, this is treated as a static string.

Related, I also typically would set docker command line options on the run command for several features I need. I may be able to get around some of this by overriding options in the docker daemon itself, but as far as passing host information I am not aware of any workaround.

**Comment by euank** [comment]
Sorry for the slow response on this; passing through host environment variables is a tricky subject.

First of all, you have a few other questions that are easier to address.

1) CPU / Memory allocation available:

Yes, running things on your instance that ECS is unaware of would cause issues there, but we want to support that use-case. Being able to override the resources ECS can use at launch makes a lot of sense to me. Good question.

2) Docker run arguments

You're correct that anything that can be passed to the docker daemon (such as the ones you reference) can be edited by using the following user-data:
```
#cloud-boothook
#!/bin/sh
echo 'OPTIONS="--dns=8.8.8.8"' > /etc/sysconfig/docker
``` 

If there are any specific docker run features you want that are not covered by our task definition, we'd be interested in hearing about them.

3) Host environment variables

To clarify a little further, when you say your environment variables are "set with cloud init" you mean you've put the desired data in /etc/environment or similar, correct?

Would being able to just specify the equivilant of docker's "--env-file" as part of a container definition fit your use case, or would that be insufficient for some reason?

**Comment by radenui** [comment]
Hello,

I re-lauch the thread as I'm interested too, and I believe my use-case is close to the one described here.

In the task definition, I can statically set anything related to the environment of the running task (domain, ...), but I also need to run my docker container with arguments depending on the host on which it is launched (host IP address, AWS instance Id, ...).

My docker run command would be something like:
```
docker run -d -p=80:80 -e HOST=$HOST myrepo/myimage java -Dmy.local.ip=$(curl 169.254.169.254/latest/meta-data/local-ipv4 2> /dev/null) -jar /path/to/my/jarfile.jar
```
And would correspond to:
```
docker run -d -p=80:80 -e HOST=myhostname myrepo/myimage java -Dmy.local.ip=10.0.0.5 -jar /path/to/my/jarfile.jar
```
Is there a way to reproduce this kind of behaviour with a task definition ?

 Thanks,

Arthur

**Comment by euank** [comment]
@radenui Currently there's not a direct way to reproduce that with a task definition.

However, if you change your image to have an entrypoint file where you resolve the above, you can get similar behavior.
For example, such an entrypoint might look like:
```bash
#!/bin/sh
export HOST=$(curl -s 169.254.169.254/latest/meta-data/local-hostname)
export LOCAL_IP=$(curl -s 169.254.169.254/latest/meta-data/local-ipv4)
exec "$@"
```
With the above entrypoint, you can reference those environment variables in your command.

I'm still leaving this issue open since I think the above is probably not the right long-term solution (especially for common things like instance ip).

Best,
Euan

**Comment by owaaa** [comment]
I had to table this temporarily but the end goal is I need consul and registrator working in ECS if I'm going to migrate our services. The consul / registrator docker containers require the $HOST and $HOSTNAME to be passed in to correctly register services and advertise the host machine IP. I was trying to avoid forking the Docker container with a solution like the entrypoint as that ties the container to AWS.

The other original blocker is on my own containers I need to set the DNS to the docker bridge via the --dns flag. I think #2 might work for that if I run the ECS agent on my own tailored machined with that setting.

**Comment by johndvorakHR** [comment]
+1

**Comment by tj** [comment]
would love to see this as well, along with most of the docker-run flags, necessary for logging & metrics unless you want to run those via ansible/terraform etc

**Comment by dgalichet** [comment]
+1

having HOST and LOCAL_IP directly defined would be much appreciated. My other option would have to link directly to the container that I wan't to access, but it's not actually possible to link to a container not defined in a Task (https://forums.aws.amazon.com/thread.jspa?threadID=179653&tstart=0)

**Comment by pschorf** [comment]
I'm also interested in this to pass database connection information

**Comment by juliencruz** [comment]
+1

Also interested in this feature as it fills in some gaps for auto-registration of containers in our environment  as well.

**Comment by rainulf** [comment]
+1 :+1: 

I would like to have this feature as well.

**Comment by pikeas** [comment]
+1, same use case as @owaaa - I want to use consul + registrator, so I need a clean way to pass host-specific config into a container.

**Comment by nexus49** [comment]
+1
Being able to pass in the host or other dynamic values specific to the host is valuable either through a environment variable or possibly through a CMD statement.

Specifically when using third party containers where you don't want to create your own version of.

**Comment by vsudilov** [comment]
+1; My use case is also described above

**Comment by ccit-spence** [comment]
@euank I have been trying you thought on the entrypoint and can't get your bash script to work.  Do you have a working example?

**Comment by euank** [comment]
@ccit-spence I put together a short example that worked for me which I've posted in [this gist](https://gist.github.com/euank/2d5a7bc74dc56f502c30).

**Comment by iainlbc** [comment]
+1, consul & registrator use case as well!

**Comment by tdensmore** [comment]
+1. Need support for docker-run flags.ECS not supporting Docker defaults is puzzling (at best). 

Why has this not been implemented yet? On March 3 the proposed workaround to ECS shortcomings was identified as "probably not the right long-term solution (especially for common things like instance ip)". Five months have passed...

**Comment by scatterbrain** [comment]
+1

**Comment by michaelgruber** [comment]
+1 Need support for `--env-file` flag.

We have variables we need to keep out of our task definitions. Currently our best option seems to be editing `/etc/syconfig/` with user data. Something like:

```bash
#!/bin/bash
yum install -y aws-cli
aws s3 cp s3://my-bucket/myenvlist /etc/myenvlist
echo 'OPTIONS="--env-file=/etc/myenvlist"' > /etc/sysconfig/docker
```

Even if that works it means having a separate cluster for each task needing an `env-file` or one `env-file` containing variables for all tasks in a cluster. An `env_file` task definition option would solve this for us. Ideally it would allow an s3 endpoint:

```json
"containerDefinitions":[
  {
    "env_file":[
      { "bucket":"my-bucket", "key":"myenvlist" }
    ]
  }
]
```

Elastic Beanstalk lets you do [something similar](http://docs.aws.amazon.com/elasticbeanstalk/latest/dg/create_deploy_docker_image.html#docker-singlecontainer-dockerrun-privaterepo) in the ` Dockerrun.aws.json` for docker private repository configuration:

```json
"Authentication":{
  "Bucket":"my-bucket",
  "Key":"mydockercfg"
},
```

**Comment by thomas83** [comment]
+1

**Comment by Lakret** [comment]
+1

**Comment by urog** [comment]
+1 to being able to execute arbitrary shell to define values for vars

**Comment by filosganga** [comment]
@urog Not all the images have a shell.

**Comment by tdensmore** [comment]
This issue has been open for close to 10 months. Does anyone know if this issue is getting any official attention (other than the +1 support from the audience) ?

**Comment by samuelkarp** [comment]
@tdensmore The +1s and use-case details are very helpful for us.

**Comment by brentonannan** [comment]
+1 - need to be able to advertise consul as running on the host IP.

**Comment by sebest** [comment]
need HOST_NAME and HOST_IP

I think that being able to pass informations available in EC2 metadata (http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html) as environment variable would solve most of the use case.

Maybe extending the JSON for the task definition currently it is like this:
```
"environment" : [
    { "name" : "HOST_NAME", "value" : "myhost" },
    { "name" : "HOST_IP", "value" : "42.42.42.42" }
]
```

adding support for a syntax like that:
```
"environment" : [
    { "name" : "HOST_NAME", "metadata" : "hostname" },
    { "name" : "HOST_IP", "metadata" : "local-ipv4" }
]
```

this would get the values from http://169.254.169.254/latest/meta-data/hostname and http://169.254.169.254/latest/meta-data/local-ipv4

**Comment by kyriesent** [comment]
+1 would love to see this. Also looking to use consul and registrator and advertise the host IP

**Comment by thomaspeitz** [comment]
+1 need this for consul (have to advertise the internal ip of the docker host to other clients, otherwise consul will announce the internal docker container ip e.g. 172..) :+1:

**Comment by thomaspeitz** [comment]
Yeah :) Found a way to get it working, without building a new Image:

My solution -> Change The Entrypoint and use wget or curl to get the ip.

Afterwards:
Entry point	["/bin/bash"]
Command	["-c","consul agent -server -advertise=$(wget -qO- 169.254.169.254/latest/meta-data/local-ipv4)"]

**Comment by owaaa** [comment]
I got consul working on ECS. I modified this https://github.com/awslabs/service-discovery-ecs-consul/blob/master/service-discovery-blog-template

**Comment by joelesalas** [comment]
:+1: to providing an easy way to retrieve the container host's hostname. This is causing an issue with Celery workers all being named the same string :)

**Comment by willejs** [comment]
+1
@samuelkarp there are 20 +1's now, can you address this now please?

**Comment by samuelkarp** [comment]
@willejs As a general rule, we don't comment on the future direction of the service, but the +1s and comments help us understand what people would like to see.  Here are the specific feature requests I see in this issue:
* Container instance hostname in environment variables
* Container instance private IP in environment variables
* Container instance private IP in command

If I'm missing anything, please let me know!

**Comment by stuartsaltzman** [comment]
+1 Would like to move to ECS but cannot until features are implemented.

**Comment by aldarund** [comment]
+1. consul+registrator use case.

**Comment by harlow** [comment]
+1 for access to private IP address in task dfn

**Comment by maclennann** [comment]
+1 setting hostname to instance id is important.

**Comment by ChrisRut** [comment]
+1 for private IP in environment variables

**Comment by msnelling** [comment]
+1 for exposing the private IP address in an environment variable. The reason being we can then write a portable container that doesn't need to detect whether it's running in EC2 to use the metadata service.

**Comment by pstorch** [comment]
+1 I need that, too.

**Comment by tdensmore** [comment]
> owaaa opened this issue on **Jan 6, 2015** · 42 comments

This issue has been open for more than a year. I am glad I stopped using ECS.

**Comment by willejs** [comment]
@samuelkarp can you actually address this?

**Comment by cbbarclay** [comment]
Thank you for the continued feedback; it's clear the community has a lot of interest in this. We are working towards supporting task definition runtime customizations like this and your use cases help prioritize our work. In the meantime, you may be able to use @euank's entrypoint example.

**Comment by devotox** [comment]
For all of you ... This image worked for me on AWS https://hub.docker.com/r/devotox/registrator/

It uses the entrypoint way mentioned above and automatically assigns an ip and the cmd is consul://consul:8500

**Comment by jmurzy** [comment]
+1. I think there is also the question of whether or not "environment variables" should be evaluated per task definition. If I want to generate a unique token (e.g etcd discovery token) that should be shared among all tasks in a given service, then I'd need another mechanism to facilitate the exchange of the token among cluster nodes unless we have a way to do it in the task declaration. And no, I do not want to create wrappers for community repos on Docker Hub just so I can proxy some variables.

**Comment by majest** [comment]
If anyone have problems setting up Consul on AWS ECS, I've prepared an image that's based on gliderlabs/docker-consul plus some short explanation what you need to do 

https://hub.docker.com/r/majest/docker-consul-ecs/

**Comment by willejs** [comment]
I am thinking I could add some kind of environment variable resolution here if the key string matches a regex
https://github.com/aws/amazon-ecs-agent/blob/master/agent/api/task.go#L199
I'm going to propose a pull request, but I don't think it will be very elegant...
Has anyone got any ideas on how this could be done in a elegant way?

This is issue no 3, and its been open for over a year now, whats that about? This is a great product, and at the start, showed great promise. The work that the AWS devs did to make this in such a short period of time looked great, however, now It appears that development of the agent has slowed down/stopped in amazon from the look of the repo commit history.

38 people have expressed interest in this, and its kind of a joke for @cbbarclay to suggest the entrypoint example, its a horrible hack, and not something you should be doing when trying to run containers. 
@cbbarclay your the principle product manager, whats going on? Do you want me to go and run a load of ec2 instances on a Mesos cluster?

**Comment by zfletcher** [comment]
Mapped ports too. When you map an ECS port dynamically (set the host port = 0) it gets chosen but the container doesn't know what the host port is, and it can't update it's service catalog/registry with that port #.

**Comment by vsudilov** [comment]
FWIW adding this as part of the instance's user data may help:

```
IP=`curl http://169.254.169.254/latest/meta-data/local-ipv4 | tr -d '\n'`
consul=<consul-server cluster location>

docker run -d \
    --name consul-agent  \
    -h $HOSTNAME \
    -p $IP:8300:8300 \
    -p $IP:8301:8301 \
    -p $IP:8301:8301/udp \
    -p $IP:8302:8302 \
    -p $IP:8302:8302/udp \
    -p $IP:8400:8400 \
    -p $IP:8500:8500 \
    --restart=always gliderlabs/consul-agent \
    -advertise $IP \
    -retry-join $consul

docker run -d \
    --name registrator \
    -v /var/run/docker.sock:/tmp/docker.sock \
    --restart=always \
    --hostname=$HOSTNAME \
    gliderlabs/registrator -ip $IP -resync 10 consul://$IP:8500
```

**Comment by mamoit** [comment]
+1

**Comment by harkathmaker** [comment]
+1

**Comment by jsh2134** [comment]
+1 I have this requirement for Consul integration as well.

**Comment by sodastsai** [comment]
+1

**Comment by TMiNus** [comment]
I have yet to review the code but I would assume that whatever we put under **hostname** in the task definition will get pass to docker in the form ///docker run -h whatever///. Is this assumption correct?

**Comment by danielnitz** [comment]
+1

**Comment by danielnitz** [comment]
A quick question for clarification on the `/etc/sysconfig/docker` approach:

Will those options be used when a task is run?

I'm using a current ECS optimized AMI (ami-67a3a90d) and modify my `/etc/sysconfig/docker` as follows (through Cloudformation):

```
OPTIONS='--add-host=docker.local:10.0.0.12'
```

Where 10.0.0.12 is the IP of the EC2 instance my container task is going to run on. As far as I understand the "add-host" option should add docker.local to every container's `/etc/hosts` running on that docker daemon. But when I inspect a container launched as a ECS task on that instance it doesn't show up.

Any idea why?

**Comment by phenomenes** [comment]
+1

**Comment by thoughtarray** [comment]
+1

**Comment by SilverGhostBS** [comment]
Hi, I think what I need is mostly related to the original request, I don't care about the HOST or IP address, but I would like to be able to just pass an arbitrary environment variable from the host to any container as is possible by using '-e NAME_OF_VAR' without any value on a regular 'docker run'. This seems currently IMPOSSIBLE using a task definition, and it is really a blocking issue.

**Comment by Bogdand** [comment]
+1

**Comment by vpack** [comment]
+1

**Comment by rjosephwright** [comment]
+1

Using a bash entrypoint wrapper is a hack at best. In my case, I cannot even do it because I want to run quay.io/coreos/etcd version 2.x, and these images do not contain bash or any other shell, only the etcd binaries. So I am stuck having to maintain my own image.

I think providing a means for users to define their own custom variables would be the most flexible and powerful way to do this, because not everyone agrees which variables are needed. I envision something like a file `/etc/ecs/ecs-vars.config`. Inside, it would contain the user's variable definitions, with the ability to run commands to get a value:

```
SOME_VAR=foo
INSTANCE_IP=`curl 169.254.169.254/latest/meta-data/local-ipv4`
```

These variables would then be available from the task definition:

```
"environment" : [
    { "name": "SOME_VAR", "custom": "SOME_VAR" },
    { "name": "INSTANCE_IP", "custom": "INSTANCE_IP" },
    { "name": "SOME_NORMAL_VAR", "value": "hello" }
]
```

**Comment by aldarund** [comment]
1.5 years since its reported.. Looks like 2-3 people working on ecs in their spare time..

**Comment by djenriquez** [comment]
Not very helpful feedback @aldarund.

My **personal** **opinion** having run several ECS clusters in production for over a year: I think its important to understand ECS' role in the containerized work space. ECS treats infrastructure as a black box, such that you tell it to run a service and it shouldn't matter where its running, but the fact that it IS running. If you're needing metadata that attaches your service to infrastructure, then you're asking for something more than what ECS offers. That service needs to run outside of ECS.

For example: the **Consul + Registrator** combo, these tools are supporting agents that work with your infrastructure. As such, should not be managed by ECS. So how does one fix this? Add it to your userdata where environment variables _can_ be referenced:
```
docker run -d --net=host \
consul agent \
-bind=$(curl http://169.254.169.254/latest/meta-data/local-ipv4) \
-retry-join=$CONSUL_SERVER_ADDRESS

docker run -d \
--name=registrator \
--net=host \
--volume=/var/run/docker.sock:/tmp/docker.sock \
gliderlabs/registrator:latest \
consul://localhost:8500
```
If you add those two commands in the same userdata you should be using to start the ECS-Agent, then you've successfully implemented the Consul + Registrator combo in ECS. 

When you scale your ECS cluster, they will all be using Consul + Registrator. When you run services in this ECS cluster, they will all be registered to Consul. Use dynamic ports, Consul/Registrator doesn't care!

Based on the needs posted above, I believe ECS has other fundamental features it needs to work on before this thread's feature request is relevant. For example, if you did want to run Consul + Registrator in ECS, it probably wouldn't do you much good anyways because ECS doesn't acknowledge instance affinities. Meaning, your cluster could be in a state such that you have 3 instances, so you tell ECS to run 3 consul/registrators. Well, you could have 3 consul agents in server "A", 2 registrators in server "B" and 1 registrator in server "C". Complete mess.

Bottom line, (opinion warning) I think its important to understand the role ECS plays and how to tie that in to be a _**single**_, but major, part of the complete solution; it doesn't do it all and I firmly believe it shouldn't . I do believe this feature IS something that would be nice, but it isn't necessary to run ECS.

**Comment by aabes** [comment]
@djenriquez  - excellent references and summary.

re: ECS and affinity - check out https://aws.amazon.com/blogs/compute/running-an-amazon-ecs-task-on-every-instance/ - which essentially uses user-data + ecs start-task to force a container to run just where you need it.

**Comment by thomask** [comment]
+1 for dynamically mapped port in an ENV variable

**Comment by genexp** [comment]
@aabes We use this approach, although with Terraform (just pulled the important bits out of the Cloudwatch script).  

https://aws.amazon.com/blogs/compute/service-discovery-via-consul-with-amazon-ecs/

I agree, Consul and Registrator should just run (via Docker if you like) on the base VM.

**Comment by dmerrick** [comment]
It's really unfortunate that this still hasn't been solved. Being able to set the hostname of a container is a basic feature, and it is very important.

**Comment by maartenvanderhoef** [comment]
@dmerrick : This is possible by setting "hostname": "your_fancy_hostname" inside the task definition.

With regards to the use of hardcoded 172.17.x.1 as host ip. With a release update Docker has once changed the docker0 address  from 172.17.42.1 to 172.17.0.1, you can use it safely by using the docker daemon option --bip 172.17.0.1/16 to fix the ip+net. This helps most Consul cases.

On EC2Metadata, AWS documentation describes task based IAM roles with optional firewalling of the ec2metadata service to disable access to the ECS related IAM role from within the containers, this would then also prevent containers to know their host ipv4 address. http://docs.aws.amazon.com/AmazonECS/latest/developerguide/ecs-dg.pdf page 199.

It would be helpful if the ecs-agent could expose the local_ipv4 address, by the same logic mentioned in the developer guide it would then be available trough 169.254.170.2. This way we can harden the containers by disallowing access to the ECS IAM Role (EC2metadata) and meanwhile provide host ipv4 for containers which need to advertise their address.

**Comment by ezeev** [comment]
+1

**Comment by andreloc** [comment]
+1

**Comment by genexp** [comment]
I wanted to share what we're doing for this..  We use Alpine Linux images,
and init all of our processes with dumb-init (
https://github.com/Yelp/dumb-init).  We run a Telegraf instance with StatsD
enabled on all of our host virtual machines.  We install this using user
data (happy to share how if anyone is interested, its not relevant to this
thread).

We develop and test via CI using Docker, so by default we don't want to use
a custom entrypoint script.  By default then, the important bits of our
Dockerfile look like this:

```Dockerfile
ENTRYPOINT ["/usr/local/bin/dumb-init", "--"]
CMD ["./appd"]
```

In our docker-compose.yml, we default an environment variable called
TELEGRAFENDPOINT to another container defined in the docker-compose file as
follows:

```docker-compose
  environment:
      - TELEGRAFENDPOINT=telegraf:8125
```

To make things work in production, we include a custom docker-entrypoint.sh
file and check it into the repo.  As I said, this isn't used locally or by
CI.

```bash
#!/usr/local/bin/dumb-init /bin/sh
set -e

export TELEGRAFENDPOINT=$(curl -s
169.254.169.254/latest/meta-data/local-ipv4):8125

exec env "$@"
```
Using ECS container definitions, we override the default entrypoint as
follows:
```javascript
    "entryPoint":["./docker-entrypoint-ecs.sh","--"],
```

This lets us keep telegraf and install it via docker-compose locally and in
CI.  It also lets us use dumb init everywhere too.


Hopefully this helps someone, it took me about 5 hours to get it all
working :)

**Comment by beeflamian** [comment]
In my app, I need to pass dynamic information to containers so that containers could be correctly configured and connected to master nodes running in our data center. However, as this thread initially stated, all envs in task definition, e.g HOST=$(hostname), would be parsed as static strings. As a workaround, we have to store all needed information in a file and mount that file to containers.

The feature request has been opened for two years with a lot of 👍 . May I humbly ask why this feature request couldn't be added to ECS agent? As docker itself is moving with a pretty fast pace, it worries me that ECS cannot be updated in time.

**Comment by rafaeldff** [comment]
+1, not every service solution is amenable to a sidecar solution such as registrator.

**Comment by genexp** [comment]
Agree. And registration itself implies consul; at least as I understand it.
Consul implies another 3 node cluster.   The whole reason to use ECS is to
eliminate the need for some of these extra bits. If they end up being
required, might as well boot up Nomad or Kub.  I think fast iteration on
ECS is pretty key.

**Comment by todd-alexion** [comment]
This issue has been open for more than two years. I am glad I stopped using ECS.

**Comment by jch254** [comment]
+1!!!

**Comment by nikhilo** [comment]
I was able to start consul in ECS with not so shabby workaround `CONSUL_BIND_INTERFACE=eth0`
Here is the full command,
```
docker run -it --rm --name=consul_ecs --net=host -e CONSUL_BIND_INTERFACE=eth0 \
consul:latest agent -server -ui -client='0.0.0.0' \
-retry-join-ec2-tag-key=Services -retry-join-ec2-tag-value=dev-consul \
-bootstrap-expect=3
```

**Comment by sijocherian** [comment]
+1 , want to use consul

**Comment by cheddesi** [comment]
Two years and still there is no neat solution to this problem! 
AWS Team, any update on this issue?

Thanks
Siva Chedde

**Comment by dmerrick** [comment]
It doesn't look like they have any intention of adding this

**Comment by danielhanold** [comment]
Here's a solution that includes creating a file containing environment variables on the Docker Host, using CloudFormation to automate the creation of this file, and then using an ENTRYPOINT script inspired by the Postgres Docker image: https://www.danielhanold.com/2017/09/set-dynamic-environment-variables-ecs-containers-using-mounted-volumes-docker-entrypoints/

**Comment by rjosephwright** [comment]
@danielhanold it isn't necessary unless you need information passed in that is unique to the underlying host, such as its IP address.

Example using CloudFormation (pardon if the syntax is not 100% correct):

```
Parameters:
  NodeEnv:
    Type: String
    AllowedValues:
      - QA
      - Staging
      - Production

Resources:
  Type: AWS::ECS::TaskDefinition
  Properties:
    ContainerDefinitions:
      - Image: my-image:v1.0
        Environment:
          - Name: NODE_ENV
            Value: !Sub NodeEnv
```

**Comment by danielhanold** [comment]
@rjosephwright I'm currently using Travis to build and upload the Docker image to ECR, then generate an updated JSON representation of the task definition, followed by an AWS CLI call to update a specific CloudFormation stack using the new task definition.

I don't know why I never thought of defining the task definition in CloudFormation itself the way you described, which sounds like a much cleaner way to manage & update containers with CFN. Thanks for the great suggestion.

**Comment by simonslater** [comment]
@danielhanold what I like about your solution is that it allows for a task definition to be used by multiple enviroments.

**Comment by adamrbennett** [comment]
Here's my use-case:
Create a task definition and run it in dev stack.
Test, iterate, etc.
When ready to promote, reuse the same task definition in prod stack.

It seems this is the intended approach, since task definitions are not cluster-specific. Furthermore, the task definition includes run-time properties, and contains things which may affect run-time behavior. It makes the most sense to promote the run-time definition, rather than just the build-time definition (i.e.: the Docker image) and then re-create, or try to match, the run-time definition with what was tested in dev.

In order to achieve this, we must be able to override the container's environment variables depending on the stack it is deployed to. This would be simple if we could use host environment variables to set values on container environment variables.

**Comment by shailesh2088** [comment]
+1
This is possible when you execute docker run on bash. I don't see why it's not possible using task definitions.

**Comment by zfletcher** [comment]
This just in....  https://aws.amazon.com/about-aws/whats-new/2017/11/amazon-ecs-allows-containers-to-directly-access-environmental-metadata/

> Now, applications running in containers managed by Amazon ECS can directly query environmental metadata such as the Docker container and Docker image name and id, the container’s networking configuration and port mapping, as well as the ECS Task’s and Container Instance Amazon Resource Name (ARN). This makes applications running on Amazon ECS environment-aware, and enables them to self-register in Service Discovery and Configuration Management solutions.

**Comment by nmeyerhans** [comment]
Thanks @zfletcher for calling that out here. Hopefully this container metadata API addresses most of the use cases captured in this issue. You had previously called out dynamically mapped host ports as one particularly interesting to you. We do support that, but be aware of #1052 for now.

I'd love to resolve this issue at this point, but I don't want to do so if there are still scenarios that aren't addressed by the container metadata feature. This issue has been open for a long time and quite a few people have contributed to it. If you've been waiting for this functionality, please give it a try and let us know if we've addressed your needs. If we haven't, please provide details. (Note that we'd also be happy to have people open new issues; it might help us keep better track of the various requests.)

**Comment by danielhanold** [comment]
@zfletcher Thanks for adding this update, but @nmeyerhans: I don't think this address or resolves the problem raised in this issue. @adamrbennett describes (what I believe to be) the most common use case that is currently not possible with ECS:
> Create a task definition and run it in dev stack.
> Test, iterate, etc.
> When ready to promote, reuse the same task definition in prod stack.

The issue is pretty close to the actual title of this issue: "Need Host environment variable resolution to pass some information to a container". Here's my use case, which is not solved by this recent addition:
* CFN Stack A that creates an ECS cluster with 1 ECS instance and sets an environment variable `ENV_NAME=DEV` on the ECS instance
* CFN Stack B that creates an ECS cluster with 1 ECS instance and sets an environment variable `ENV_NAME=PROD` on the ECS instance
* Task Definition X that is running a container on Stack A, which uses `ENV_NAME` from the ECS Cluster created by Stack A => `ENV_NAME = DEV`
* Once tested successfully, the _very same_ task definition X is then promoted to Stack B and the container uses the `ENV_NAME` from the ECS Cluster created by Stack B => `ENV_NAME = PROD`

Other than this use case, it's frustrating that we can't use variables for the container environment variables in the task definition and are limited to the hardcoded values set in a task definition.

**Comment by SilverGhostBS** [comment]
@danielhanold Thanks, that's EXACTLY my use case.

What's more frustrating is that this is allow through command line on the vanilla docker, just put -e VAR_NAME and it will passthrough the env var from host to container.

But the task definition schema absolutely insists on us setting a value for the parameter, whereas just allowing to omit it might work.

**Comment by pikeas** [comment]
Agree with @danielhanold!

It doesn't make sense to have separate task definitions for dev/staging/prod. I'd like to define my needed env vars once in the task definition, then populate values in each environment's service definition.

**Comment by simonslater** [comment]
I agree with @danielhanild too!

**Comment by mceg** [comment]
The workaround that I am using: 
```
{
    "name": "kafka",
    "image": "confluentinc/cp-kafka:4.0.0",
    "essential": true,
    "memoryReservation": 2048,
    "command": [
      "/bin/bash",
      "-c",
      "export KAFKA_ADVERTISED_LISTENERS=PLAINTEXT://$(hostname -i):9092; /etc/confluent/docker/run"
    ],
    "environment": [
      {
        "name": "KAFKA_ZOOKEEPER_CONNECT",
        "value": "..."
      }
    ],
...
}
```

a little modification of entrypoint approach but without the need to manage your own containers, and without running containers from user data scripts, just registering them as services as it should be.

**Comment by pb0101** [comment]
The real problem comes when you are using third part docker images that you can't modify and add another ENTRYPOINT. 
Is there any other solution besides modifying ENTRYPOINT

**Comment by zfletcher** [comment]
You can usually do `docker run --entrypoint "/bin/bash" ...` or similar.

**Comment by rash805115** [comment]
I faced this problem today and after hours of research I found this thread. Wow, the discussion here is very informative and it helped me find a solution for myself.

As suggested in this thread I modified my Dockerfile to execute a script at entrypoint and in that script I made the call to metadata service to get LOCAL_IP.
Please refer to [my StackOverflow post](https://stackoverflow.com/a/53245823/1834562) for more details.

**Dockerfile**
```
FROM gliderlabs/registrator:v7

COPY entrypoint.sh /entrypoint.sh
RUN chmod +x /entrypoint.sh && \
    apk update && \
    apk add curl

ENTRYPOINT ["/entrypoint.sh"]
```

**entrypoint.sh**
```
#!/bin/sh

LOCAL_IP=$(curl -s --connect-timeout 3 169.254.169.254/latest/meta-data/local-ipv4)
if [ $? -eq 0 ]; then
  /bin/registrator -ip $LOCAL_IP $@
else
  /bin/registrator $@
fi
```

I hope this issue is fixed the right way, but until then this solution would suffice for me.

**Comment by chaitan94** [comment]
The entrypoint solution suggested by multiple people here definitely works, but really just feels like a hacky way.

This issue has been open for a long time now, is there any specific reason why there has been no elegant solution for this so far? I'm really curious to know what are the hurdles here. I'll be even glad to contribute myself if someone can explain what's going wrong here.

**Comment by mfortin** [comment]
Can't the labels be leveraged ? The agent can define them, and they can be retrieved using the metadata URI. I am looking at container instance tags to retrieve the information somehow, haven't found a clean solution yet.

**Comment by stefanjauker** [comment]
+1

**Comment by xyvark** [comment]
Just want to add my 2 cents here as this is something that would be really useful for us if it worked the same way it works in docker: `-e ENV_VAR` without value uses the value from the host. Would it work if we could set environment variables without values in the task definition?

**Comment by bvanderveen** [comment]
+1.

What I want to do is to build my set of containers, build a task def that orchestrates that set of containers, and then launch that taskdef into a cluster with different environment variables (say, to distinguish between dev/prod). 

Seems my only choices are to a) maintain separate dev/prod task defs, where the only difference is some env var that indicates dev vs. prod or b) bake the fact of being dev or prod into the containers (which seems completely absurd). 

To be honest, I feel like I've been sold a bill of goods, but I'm far enough into the finger trap and Stockholm syndrome that I have no choice but to press on with my 3 weeks of sunk cost and my dreams for a better world.

And this is after a buddy of mine took his startup's infrastructure down by deploying during a scale-out event two days before getting a Letter of Intent from a megacorp learning "the hard way" that ECS Blue/Green deploys don't support Auto-Scaling. 

Guys. This is borderline bad-faith.
